src/runtime/virtcontainers/utils/utils.go

```diff
@@ -306,6 +306,36 @@ func ConvertAddressFamily(family int32) pbTypes.IPFamily {
 	}
 }
 
+func waitProcessUsingWaitLoop(pid int, timeoutSecs uint, logger *logrus.Entry) bool {
+	secs := time.Duration(timeoutSecs)
+	timeout := time.After(secs * time.Second)
+
+	for {
+		// Check if the process is running periodically to avoid a busy loop
+		// "A watched pot never boils" and an unwaited-for process never appears to die!
+		waitedPid, err := syscall.Wait4(pid, nil, syscall.WNOHANG, nil)
+
+		if waitedPid == pid && err == nil {
+			return false
+		}
+
+		if err := syscall.Kill(pid, syscall.Signal(0)); err != nil {
```

Comment:
This may be related to the "watched pot never dies" comment above, but why do you need both `wait4` and `kill(sig, 0)` in the same loop?

Reply:
`Wait4` call will succeed for a child process (and reap the child also). Using only `kill(0, pid)` will return nil as long the process isn't reaped, so without Wait4 the function will timeout. For non-child processes, using wait4 will fail. kill(0, pid) will fail with ESRCH error code once the process is reaped by `init` and the function will return false.

Reply:
@alex-matei If you have time to rephrase the comment to explain this logic. The wording of the comment above seems even more concise than @gkurz's, so it might be the content of the comment.